File: infra/io/binary/src/main/java/dev.mzarnowski.io.binary.v5/IntEncoding.java
package dev.mzarnowski.io.binary.v5;

import org.jetbrains.annotations.NotNull;

public final class IntEncoding implements Encoding.OfInt {
    private final int size;

    public IntEncoding(int size) {
        this.size = size;
    }


    @Override
    public int write(@NotNull Buffer buffer, int offset, int value) {
        switch (size) {
            case 4:
                buffer.write(offset + 3, value >> 24);
            case 3:
                buffer.write(offset + 2, value >> 16);
            case 2:
                buffer.write(offset + 1, value >> 8);
            case 1:
                buffer.write(offset, value);
        }

        return size;
    }
}

File: infra/io/binary/src/main/kotlin/dev/mzarnowski/io/binary/encoding/IntEncoding.kt
//package dev.mzarnowski.io.binary.encoding
//
//class IntEncoding internal constructor(val bytes: Int) : Encoding.OfInt {
//    override fun write(value: Int, buffer: ByteArray, offset: Int) {
//        var shift = 8 * (bytes - 1)
//        repeat(bytes) {
//            buffer[offset + it] = (value ushr shift).toByte()
//            shift -= 8
//        }
//    }
//
//    override fun read(buffer: ByteArray, offset: Int): Int {
//        var value = 0
//        var shift = 8 * (bytes - 1)
//        repeat(bytes) {
//            value += buffer[offset + it].toInt() shl shift
//            shift -= 8
//        }
//        return value
//    }
//
//    companion object {
//        val OneByte = IntEncoding(1)
//        val TwoBytes = IntEncoding(2)
//        val ThreeBytes = IntEncoding(3)
//        val FourBytes = IntEncoding(4)
//    }
//}